tools/rewriter/GenCallAsm.cpp

```diff
@@ -11,125 +12,168 @@
-                                          "xmm4", "xmm5", "xmm6", "xmm7"};
+const std::vector<const char *> arm_float_reg_order = {"q0", "q1", "q2", "q3",
+                                                       "q4", "q5", "q6", "q7"};
 
-const std::array<const char *, 2> int_ret_reg_order = {"rax", "rdx"};
+const std::vector<const char *> x86_int_ret_reg_order = {"rax", "rdx"};
+// return and parameter registers are the same on AArch64
+const std::vector<const char *> arm_int_ret_reg_order = arm_int_param_reg_order;
 
 const std::array<const char *, 3> cabi_arg_kind_names = {"int", "float", "mem"};
 
 // rsp and rbp are also preserved registers, but we handle them separately from
 // these since they're the stack and frame pointers
-const std::array<const char *, 5> preserved_registers = {"rbx", "r12", "r13",
-                                                         "r14", "r15"};
+const std::array<const char *, 5> x86_preserved_registers = {"rbx", "r12", "r13",
+                                                             "r14", "r15"};
+
+const std::array<const char *, 10> aarch64_preserved_registers = {"x19", "x20", "x21",
+                                                                  "x22", "x23", "x24",
+                                                                  "x25", "x26", "x27",
+                                                                  "x28"};
 
 /// Compute abi locations for parameters of a C-abi function, given its sequence
 /// of argument kinds.
-std::vector<ParamLocation> param_locations(const CAbiSignature &func) {
-  std::vector<ParamLocation> locs = {};
+std::vector<ArgLocation> allocate_param_locations(const AbiSignature &func, Arch arch, size_t *stack_arg_size_out) {
+  const auto &int_param_reg_order = (arch == Arch::X86) ? x86_int_param_reg_order : arm_int_param_reg_order;
+  const auto &float_reg_order = (arch == Arch::X86) ? x86_float_reg_order : arm_float_reg_order;
+  std::vector<ArgLocation> args;
   size_t ints_used = 0;
-  // if the return is in memory, the first integer argument is the location to
-  // write the return value
   size_t memory_return_slots =
       std::count_if(func.ret.begin(), func.ret.end(),
-                    [](auto &x) { return x == CAbiArgKind::Memory; });
+                    [](auto &x) { return x.is_stack() || x.is_indirect(); });
   if (memory_return_slots > 0) {
-    locs.push_back(ParamLocation::Register(int_param_reg_order[0]));
-    ints_used++;
+    if (arch == Arch::X86) {
+      // if the return is in memory, the first integer argument is the location to
+      // write the return value
+      auto ptr_arg = ArgLocation::Register(ArgLocation::Kind::Integral, 8);
+      ptr_arg.allocate_reg(int_param_reg_order[ints_used]);
+      args.push_back(ptr_arg);
+      ints_used += 1;
+    } else if (arch == Arch::Aarch64) {
+      // memory return goes in x8
+      auto ptr_arg = ArgLocation::Register(ArgLocation::Kind::Integral, 8);
```

Comment:
I'm interpreting this as the return value goes in an 8 byte slot on the stack which x8 points to. Why is this 8?

Reply:
This is the register (x8) holding the address of the stack return slot. I'll clarify.